### ISSUE TITLE
Some questions about bismark_methylation_extractor

### ISSUE THREAD

**Issue by sunshine-lp0** [FelixKrueger/Bismark#205]
Dear Felix,
    I'm writing a workflow about DNA methylation. But I got confused while dealing with bismark_methylation_extractor. 

**MY CODE:**
```
#Step1
Bismark --non_directional –unmapped –fastq --phred33-quals
--genome_folder /PATH/
--bowtie2 --path_to_bowtie /PATH/
-1 hEC_WT_1_R1.fastq.gz 
-2 hEC_WT_1_R2.fastq.gz 
--output_dir /PATH/
--temp_dir /PATH/
--parallel 5

#Step2
deduplicate_bismark --bam --paired /PATH/*bismark_bt2_pe.bam 
--samtools_path /PATH/ 
--output_dir /PATH/

# Step3
bismark_methylation_extractor --p --gzip --comprehensive 
--bedGraph --cytosine_report --zero_based *bismark_bt2_pe.deduplicated.bam 
--genome_folder /PATH/
--output /PATH/
--multicore 10 

#Step4
bismark2bedGraph --buffer 10G --CX -o bismark2bedGraph_CpG_context CpG_context_*_bismark_bt2_pe.deduplicated.txt.gz
bismark2bedGraph --buffer 10G --CX -o bismark2bedGraph_CHH_context CHH_context_*_R1_bismark_bt2_pe.deduplicated.txt.gz
bismark2bedGraph --buffer 10G --CX -o bismark2bedGraph_CHG_context CHG_context_*_bismark_bt2_pe.deduplicated.txt.gz

#Step5
coverage2cytosine --gzip --genome /PATH/ -o coverage2cytosine_CpG_context bismark2bedGraph_CpG_context.gz.bismark.cov.gz
coverage2cytosine --CX --gzip --genome /PATH/ -o coverage2cytosine_CHH_context bismark2bedGraph_CHH_context.gz.bismark.cov.gz
coverage2cytosine --CX --gzip --genome /PATH/ -o coverage2cytosine_CHG_context bismark2bedGraph_CHG_context.gz.bismark.cov.gz
```

**### Question 1:**
After Step3, there are 3 kinds of output files. My genome is **Ensembl GRCh38**-- one-based coordinates. 

**Question 1.1** I don't know which output file is right. 
**Question 1.2** Why *.bedGraph.gz file and .bismark.cov.gz file don't have context information?
**Question 1.3** *.CpG_report.txt.gz file has context information. But Why is there such a situation: both `<count methylated> `and `<count non-methylated>` are 0 ? 
**Question 1.4**: I tried Step4 and Step5 to solve this problem. There is also a situation `<count methylated>=0 <count non-methylated>`0 in coverage2cytosine_CHG_context.CpG_report.txt.gz.

```
output file 1: *bismark_bt2_pe.deduplicated.bedGraph.gz
track type=bedGraph
17      60054   60055   100
17      60141   60142   100
17      60168   60169   100
17      62594   62595   100
17      62603   62604   100
17      62606   62607   100
17      62663   62664   0
17      62734   62735   100
17      62753   62754   100

output file 2: *bismark_bt2_pe.deduplicated.bismark.cov.gz
17      60055   60055   100     2       0
17      60142   60142   100     2       0
17      60169   60169   100     2       0
17      62595   62595   100     1       0
17      62604   62604   100     1       0
17      62607   62607   100     1       0
17      62664   62664   0       0       1
17      62735   62735   100     1       0
17      62754   62754   100     1       0

output file 3: *bismark_bt2_pe.deduplicated.CpG_report.txt.gz
17      60053   +       0       0       CG      CGC
17      60054   -       2       0       CG      CGG
17      60140   +       0       0       CG      CGC
17      60141   -       2       0       CG      CGG
17      60167   +       0       0       CG      CGC
17      60168   -       2       0       CG      CGT
17      60200   +       0       0       CG      CGA
17      60201   -       0       0       CG      CGG
17      60209   +       0       0       CG      CGT
17      60210   -       0       0       CG      CGT

```
**### Question 2:**
There are 3 kinds of output file format described in bismark_methylation_extractor --help. Why they apply to different coordinates? BedGraph output(0-based start coords, 1-based end coords); coverage output(1-based genomic coords; zero-based half-open coordinates); genome-wide cytosine methylation output(?). Which options should I choose when I use different genome(Ensembl, UCSC, or NCBI)?

**### Question 3:**
Can bismark report a file including all information(below)? And I just want to get CG information.
`<chromosome>,<position>,<base:C/G?>,<strand>,<total count>,<count methylated>,<count non-methylated>,<methylation percentage>,<c-context>,<trinucleotide context>`

**### Question 4:**
How do you think if I run paired-end alignment first and then run single-end alignment using unmapped read? Is it more correct? If i can do it. Which step result should be merge? 

Best best best wishes,
Anita

**Comment by FelixKrueger** [comment]
That's a lot of text... The first thing I noticed was that the CpG context processing in **Step 4** and **Step 5** is redundant as you already call both of them at the **Step 3** with the options `--bedGraph` and `--cytosine_report`. Just lose them. 

**Q1.1**: I don't quite understand the question here. The output files are all 'right', they are just in different formats.
**Q1.2**: That's just how it is... You know the context because of the files you use as input. If you require context information you can run `coverage2cytosine`.
**Q1.3**: The CpG_report.txt.gz file is genome based, so every single position in the genome will be present. This is so that you can have a unified matrix for several samples even if some positions are not covered. You are welcome to filter out 0/0 positions later on using `sed/awk` or some filtering script.
**Q1.4**: see Q1.3

**Q2**:  You should choose the coords that you are using downstream. Generally, Ensembl uses 1-based coords, and this is the Bismark default. UCSC generally uses 0-based coords. The bedGraph format is defined, and we therefore have to adhere to their standards. So if you work with Ensembl I would drop the option `--zero_based`.

**Q3**: All the information is there (probably in the C-report), but you might have to re-arrange them using a custom script or `sed/awk` command.

**Q4**: You can do this but we tend to not bother doing this. In a proper good experiment the single-end data will not bring much extra to the table. If you were to do it you would need to do the alignment, deduplication and methylation extraction separately, and you could merge the data for the `bismark2bedGraph` step, e.g.: `bismark2bedGraph -o bismark2bedGraph_CpG_context CpG*`

Hope this helps?

**Comment by sunshine-lp0** [comment]
wow！I almost understand. Thanks a million!!!

**Comment by sunshine-lp0** [comment]
Oh, I know. `The CpG_report.txt.gz file is genome based, so every single position in the genome will be present.`  You mean **every single C position** in the genome will be present.

**Comment by FelixKrueger** [comment]
Yes, in the context of `CpG` it will be every single C position in CpG context, and if you specify `--CX` it will indeed be **every single C position** (both top and bottom strands).

**Comment by sunshine-lp0** [comment]
For the data of paired-end sequencing, can I use ` bismark --single_end sample_R1.fastp.barcode.fastq.gz,sample_R2.fastp.barcode.fastq.gz --sam-no-hd`   to merge Reads 1 and Reads 2 ?

**Comment by FelixKrueger** [comment]
No that won't work, because R1 and R2 require different alignment modes when aligned in SE mode:

R1: `bismark --genome .. sample_R1.fastp.barcode.fastq.gz` 
R2: `bismark --genome .. --pbat sample_R2.fastp.barcode.fastq.gz`

then you can merge the BAM files with `samtools cat` or `samtools merge. Is that for reads that do not align in PE mode?

**Comment by sunshine-lp0** [comment]
Q1: I don't understand why add `--pbat` when aligning R2? Our method is based on RRBS. And i think i should use `--non_directional` ?
Q2: Our lab always align PE mode first and then align SE mode using unmapped reads(I asked you last time). But for the new workflow, i just wanna use SE mode for R1 and R2 respectively, like you told me last time.

**Comment by sunshine-lp0** [comment]
Or maybe I can :
R1: bismark --genome .. --single-end --non-directional sample_R1.fastp.barcode.fastq.gz
R2: bismark --genome .. --single-end --non-directional sample_R1.fastp.barcode.fastq.gz
and then merge the BAM files with samtools merge ?

**Comment by sunshine-lp0** [comment]
So you means i can only use PE mode for my paired-end data?

**Comment by FelixKrueger** [comment]
Well, paired-end data is paired-end data. If you treat it as single end data you will get (somewhat) different results because the data is not meant to be treated as 2 independent single-end reads.

As I have already mentioned in my very first reply I would do a paired-end alignment with your paired-end data, and probably not bother about single-end aligning unaligned reads because it is normally negligible.

If you still want to do PE first, and then SE align the unmapped reads you would need to align R1 in default (directional) and R2 in `--pbat` mode. This sequential mode would have the advantage that you can extract the methylation calls from the PE file so that you discard overlapping read portions (which you can't do if you run SE alignments). If you take a look at this guide https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/RRBS_Guide.pdf it might explain the issue a bit better.

**Comment by FelixKrueger** [comment]
You're very welcome!

**Comment by FelixKrueger** [comment]
To Q1: I think here I have to agree with you, I updated the RRBS Guide accordingly.

To Q2: Yes I think your understanding there is correct. Trim Galore trims paired-end files in exactly that format when you specify the option `--rrbs`.

To Q3: In the past we have tried to extract the methylation state from these artificial positions (instead of just removing them). The outcomes were somewhat mixed though, depending on the person who prepared the library etc, so in the end we stopped doing that. Spike-ins are in theory useful, however in practice we see that the results can also vary hugely. I believe we had everything from hardly any spike-in reads to > 90% of the library consisting of spike-in, and from a really nice control to being ~50% methylated in all contexts even though the non-CG methylation of the samples were at 0.5% methylation. Overall I am not surprised that most people tend not to use spike-ins at all... (at least this is my personal impression).
